pkg/lightning/backend/local/local.go

```diff
@@ -2090,150 +2588,263 @@ const (
+}
+
+func (h *sstIterHeap) Len() int {
+	return len(h.iters)
+}
+
+func (h *sstIterHeap) Less(i, j int) bool {
+	return bytes.Compare(h.iters[i].key, h.iters[j].key) < 0
+}
+
+func (h *sstIterHeap) Swap(i, j int) {
+	h.iters[i], h.iters[j] = h.iters[j], h.iters[i]
+}
+
+func (h *sstIterHeap) Push(x interface{}) {
+	h.iters = append(h.iters, x.(*sstIter))
 }
 
-// kvMemCache is an array of KV pairs. It also keep tracks of the total KV size and whether the array is already sorted.
-type kvMemCache struct {
-	kvs       []common.KvPair
-	totalSize int64
-	notSorted bool // record "not sorted" instead of "sorted" so that the zero value is correct.
+func (h *sstIterHeap) Pop() interface{} {
+	item := h.iters[len(h.iters)-1]
+	h.iters = h.iters[:len(h.iters)-1]
+	return item
 }
 
-// append more KV pairs to the kvMemCache.
-func (m *kvMemCache) append(kvs []common.KvPair) {
-	if !m.notSorted {
-		var lastKey []byte
-		if len(m.kvs) > 0 {
-			lastKey = m.kvs[len(m.kvs)-1].Key
+func (h *sstIterHeap) Next() ([]byte, []byte, error) {
+	for {
+		if len(h.iters) == 0 {
+			return nil, nil, nil
 		}
-		for _, kv := range kvs {
-			if bytes.Compare(kv.Key, lastKey) <= 0 {
-				m.notSorted = true
-				break
+
+		iter := h.iters[0]
+		if iter.valid {
+			iter.valid = false
+			return iter.key, iter.val, iter.iter.Error()
+		}
+
+		var k *pebble.InternalKey
+		k, iter.val = iter.iter.Next()
+		if k != nil {
+			iter.key = k.UserKey
+			iter.valid = true
+			heap.Fix(h, 0)
+		} else {
+			err := iter.Close()
+			heap.Remove(h, 0)
+			if err != nil {
+				return nil, nil, errors.Trace(err)
 			}
-			lastKey = kv.Key
 		}
 	}
+}
 
-	m.kvs = append(m.kvs, kvs...)
-	for _, kv := range kvs {
-		m.totalSize += int64(len(kv.Key)) + int64(len(kv.Val))
-	}
+// sstIngester is a interface used to merge and ingest SST files.
+// it's a interface mainly used for test convenience
+type sstIngester interface {
+	mergeSSTs(metas []*sstMeta, dir string) (*sstMeta, error)
+	ingest([]*sstMeta) error
 }
 
-// sort ensures the content is actually sorted.
-func (m *kvMemCache) sort() {
-	if m.notSorted {
-		sort.Slice(m.kvs, func(i, j int) bool { return bytes.Compare(m.kvs[i].Key, m.kvs[j].Key) < 0 })
-		m.notSorted = false
+type dbSSTIngester struct {
+	e *File
+}
+
+func (i dbSSTIngester) mergeSSTs(metas []*sstMeta, dir string) (*sstMeta, error) {
+	if len(metas) == 0 {
+		return nil, errors.New("sst metas is empty")
+	} else if len(metas) == 1 {
+		return metas[0], nil
 	}
+
+	start := time.Now()
+	newMeta := &sstMeta{}
+	mergeIter := &sstIterHeap{
+		iters: make([]*sstIter, 0, len(metas)),
+	}
+	for _, p := range metas {
+		f, err := os.Open(p.path)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		reader, err := sstable.NewReader(f, sstable.ReaderOptions{})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		iter, err := reader.NewIter([]byte{'t'}, []byte{'u'})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		key, val := iter.SeekGE([]byte{'t'})
```

Comment:
🤔 seems unnecessary to restrict the range. we don't store non-`t` prefixed stuff in the SST.

Reply:
@glorv (line 2748 is included)

Reply:
I supposed the `NewIter` must receive a not empty bound, but found we can just pass `nil, nil` instead. 😂